[V1][Spec Decode] Eagle Model loading

LiuXiaoxuanPKU opened the pull request:
Task 1 of #15901

Some limitations:
1. only tested with single GPU
2. only works with eager mode for both target and draft models, need to check compatibility with torch compile
3. only support llama models

How to run this PR:
`python examples/offline_inference/eagle.py`
Ignore the metrics used/printed in that file.

LiuXiaoxuanPKU committed:
register model

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
fix config and init

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
torch compile

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
remove lambda, add prefix

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
runnable v1

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
change the way of getting vllm_config, cleanup

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
minor

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

markmc commented:
I'm sure this has been discussed elsewhere, but why introduce a V1-specific model (`LlamaForCausalLMEagle`) instead of using the existing model (`EagleModel`) ?

e.g. it looked like @luyuzhe111 expected `EagleModel` could be used in V1 with DeepSeek MTP weights?

markmc commented on examples/offline_inference/eagle.py:
```diff
@@ -70,6 +70,7 @@
     max_num_seqs=args.max_num_seqs,
     gpu_memory_utilization=0.8,
     speculative_config={
+        "method": "eagle",
```
This is only need to paper over a bug in `_is_v1_supported_oracle()` - we're calling this before the auto-detection of spec decoding method happens in `SpeculativeConfig.__post_init__()`:

```
		if "eagle-" in self.draft_model_config.model.lower():
                    self.method = "eagle"
```

markmc replied:
```suggestion
        "method": "eagle",  # FIXME: should not be needed, but V1 oracle happens before auto-detection
```

luyuzhe111 replied:
Hi Mark @markmc ,

Not sure about the best practice but personally I think we should directly specify the method. I think this auto detection

>"eagle-" in self.draft_model_config.model.lower()

is actually not reliable since this assumes the draft model **directory** contains this "eagle-" string, which may not be the case?

ekagra-ranjan commented on vllm/v1/spec_decode/eagle.py:
```diff
@@ -176,26 +182,27 @@ def prepare_inputs(
+            self.vllm_config.speculative_config.draft_model_config
+        # FIXME(lily): This does not handle with distributed inference.
+        target_device = self.vllm_config.device_config.device
+        # We need to set the vllm_config here to register attention
+        # layers in the forward context.
```
Do we need to call load_model() from the __init__ so that this function runs and the attn layer are registered?

LiuXiaoxuanPKU replied:
Could you elaborate a bit on which `load_model` are you talking about?

ekagra-ranjan replied:
The one in this file `spec_decode/eagle.py` here: https://github.com/vllm-project/vllm/pull/16035/files/59ee450306d3d719f78ad60c77ba9b739bc5cb11#diff-a4809a837fbf535a8f0999b11087a53ec1c53948b50c0a1fe64396bc86de9461R184

ekagra-ranjan commented on vllm/v1/spec_decode/eagle.py:
```diff
@@ -176,26 +182,27 @@ def prepare_inputs(
+        # We need to set the vllm_config here to register attention
+        # layers in the forward context.
+        with set_default_torch_dtype(
+                draft_model_config.dtype), set_current_vllm_config(
```
I didnt get this part which says that setting the current vllm config will lead to registering attn layers. Can you pls share more?

My understanding is that, we did not change any self.vllm_config in this function and the attn are registered when the model is init such that it gets saved in [static_forward_context](https://github.com/vllm-project/vllm/blob/4708f13a9c8758d08cd45df200d63aa97762e873/vllm/attention/layer.py#L148) which is then used during `bind_kv_cache()`

LiuXiaoxuanPKU replied:
Which 'load_model' are you referring to? This [one](https://github.com/vllm-project/vllm/blob/2755c34a8fef8d0c1b0292a7ddae45d8f92f070c/vllm/model_executor/model_loader/loader.py#L446)?

ekagra-ranjan replied:
The one in spec_decode/eagle.py here: https://github.com/vllm-project/vllm/pull/16035/files/59ee450306d3d719f78ad60c77ba9b739bc5cb11#diff-a4809a837fbf535a8f0999b11087a53ec1c53948b50c0a1fe64396bc86de9461R184

LiuXiaoxuanPKU committed:
fix weights loading

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
fix layernorm and fix cudagraph

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU commented:
Hi folks, thanks for all the comments so far, I am double checking the correctness and have not started fixed the comments. Some updates:
1. I checked the compatibility with torch.compile, it should work. Concretely, the target model is running with torch.compile cudagraph, but the head is running in eager mode.
2. I fixed some correctness bugs in the head definition.
3. I benchmarked the performance of the example in [here](https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/eagle.py) on H100 (I just changed the eagle model to `yuhuili/EAGLE-LLaMA3-Instruct-8B`, which is downloaded from huggingface directly). The numbers mean request latency:

| Eagle (k=1) | Eagle (k=2)  | Eagle (k=3) | Eagle (k=4) | w/o Eagle|
|---|---|---|---|---|
|1.49 | 1.39 | 1.43 | 1.47 | 1.92 |

Please start review and check correctness. cc @luyuzhe111 @WoosukKwon.

ekagra-ranjan commented:
@LiuXiaoxuanPKU Good results! 

I was wondering how are we able to run EAGLE give Task 2, 3 are not done in https://github.com/vllm-project/vllm/issues/15901 ? What are the implications/assumptions of these wrt to the results shared in this PR?

LiuXiaoxuanPKU commented:
> @LiuXiaoxuanPKU Good results!
> 
> I was wondering how are we able to run EAGLE give Task 2, 3 are in #15901 are WIP ? What are the implications/assumptions of these wrt to the results shared in this PR?

Thanks for asking:
1. task 2 is about allocating KV cache. When batch size is small (there is enough kv cache), the current implement should not cause errors such as overwriting KV cache of some other requests.
2. task 3 is mainly for standard sampling. I 'm benchmarking greedy sampling here.

LiuXiaoxuanPKU committed:
example

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

ekagra-ranjan commented:
@LiuXiaoxuanPKU - are results on gsm8k? If possible can you run them on MTBench so that we can compare the results with SGL and identify gaps? https://docs.google.com/document/d/18ETJLsnxR88Qq3VDk5Mq-Hb7vuE9o3VNZ-hhz-OqAXk/edit?usp=sharing

luyuzhe111 commented:
@LiuXiaoxuanPKU Hi Lily, when running with `VLLM_USE_V1=1`, I am getting errors from attention backend. Wondering if there is additional env variables that I need to set?

v-lmn commented:
Why is your test using the v0 interface?

LiuXiaoxuanPKU committed:
revert example and test

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
Merge branch 'main' into eagle-weight

LiuXiaoxuanPKU committed:
merge

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

luyuzhe111 commented:
can confirm the example runs without errors. to test correctness we need to collect acceptance rates, which is probably beyond the scope of this PR.

ekagra-ranjan commented:
@luyuzhe111 - metrics are already [here](https://github.com/vllm-project/vllm/pull/15151) so we should run them in this PR since we already benchmarked the latency?

WoosukKwon commented on vllm/model_executor/models/llama_eagle.py:
```diff
@@ -0,0 +1,151 @@
+        return loaded_params
+
+
+class LlamaForCausalLMEagle(LlamaForCausalLM):
```
nit:
```suggestion
class EagleLlamaForCausalLM(LlamaForCausalLM):
```

WoosukKwon commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1191,11 +1191,15 @@ def execute_model(
+                target_hidden_states = hidden_states[:num_scheduled_tokens, :]
                 target_slot_mapping = attn_metadata.slot_mapping
                 cu_num_tokens = attn_metadata.query_start_loc
+                token_indices = None
```
Do we need this?

luyuzhe111 commented:
> @luyuzhe111 - metrics are already [here](https://github.com/vllm-project/vllm/pull/15151) so we should run them in this PR since we already benchmarked the latency? The latency depends on how efficiently eagle was implemented but the Acc Length would be implementation independent mostly and would reflect the correctness

Hi @ekagra-ranjan, if I don't think spec dec metric is hooked up to the LLM API yet?

LiuXiaoxuanPKU committed:
fix comments

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU commented on vllm/model_executor/models/llama_eagle.py:
```diff
@@ -0,0 +1,151 @@
+        super().__init__()
+        self.config = model_config.hf_config
+        self.vocab_size = self.config.vocab_size
+        self.embed_tokens = VocabParallelEmbedding(
```
yes, it's the same, it's just the eagle head also includes the embed_tokens weights, so we just load it here. We can also set it to the embed_token of target model. I double check the weights of the two, they are the same.

WoosukKwon replied:
Thanks for checking it. I'm ok with having it for now, but maybe worth a comment?

ekagra-ranjan replied:
If we don't set it to the embed_token of target model then we are loading the vocab twice which take more GPU memory for larger vocab?

WoosukKwon replied:
@ekagra-ranjan Yes, for now. We can share the embedding when PP size is 1, but I was thinking that we can do this in a followup PR as an optimization.

ekagra-ranjan replied:
Fixed by: https://github.com/vllm-project/vllm/pull/17326

WoosukKwon reviewed (commented):
LGTM. Can you add an e2e test (or pull an existing one from v0)?

LiuXiaoxuanPKU committed:
add tests

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
less strict

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
fix registry test

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

LiuXiaoxuanPKU committed:
minor

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

ekagra-ranjan commented:
@LiuXiaoxuanPKU 

I benchmarked this PR on MTBench using the `lmsys/sglang-EAGLE-LLaMA3-Instruct-8B` so that we can compare with the [SGL benchmark](https://docs.google.com/document/d/18ETJLsnxR88Qq3VDk5Mq-Hb7vuE9o3VNZ-hhz-OqAXk/edit?usp=sharing) I did sometime back to help us get the direction.

I wired up the SD metrics in Scheduler in EngineCoreOutputs so that we can test the correctness with the Accept Length. Currently, the SD metrics in Scheduler gets reinitialized every engine step and is not aggregated so I fix that. [Here](https://github.com/ekagra-ranjan/vllm/pull/1) is a dummy PR which has the benchmarking script and changes I did to get the below results. Pls lmk if something is incorrect in my changes. I can also raise a PR here if that makes sense with some extra steps.

Here is the cmd used
```
VLLM_USE_V1=1 python examples/offline_inference/eagle.py --dataset="../data/mt_bench/question.jsonl" --num_spec_tokens 4 --max_num_seqs 1
```
`num_spec_tokens` is 2 or 4. 

* vanilla, [02:08<00:00,  1.61s/it, est. speed input: 47.09 toks/s, output: 128.64 toks/s]
* k=2: [01:34<00:00,  1.18s/it, est. speed input: 64.12 toks/s, output: 175.49 toks/s] **Accpt Len: 1.89**
* k=4: [01:40<00:00,  1.25s/it, est. speed input: 60.52 toks/s, output: 165.65 toks/s] **Accpt Len: 2.08**

k=2 is 36% faster and k=4 is 28% faster than vanilla. 

Compared to the SGL bench:
* eagle gains are nearby
* the absolute throughput of vllm is 14% lower than SGL
* Accept Length (AL) is certainly lower: 2.08 (vLLM) vs 2.44 (SGL). 
    * vllm AL formula is [here](https://github.com/ekagra-ranjan/vllm/pull/1#discussion_r2036110687)
    * SGL AL formula is [here](https://github.com/sgl-project/sglang/blob/456b008bd8f130a79856dbadbae0e44aeaad3bfd/benchmark/mtbench/bench_sglang_eagle.py#L101)
 
I feel both the formulas are similar and that shouldnt be the cause of differences in the metric but pls lmk if I something is missing.

WoosukKwon commented:
@ekagra-ranjan This PR itself is not enough to support eagle correctly. We need to handle the draft model's KV cache properly.

luyuzhe111 commented:
@ekagra-ranjan Hi Ekagra, 

Thanks for sharing the benchmarking results! I can confirm the acceptance length you collected for vllm should be accurate, as I aggregate acceptance length from request-level, per-step acceptance counts in this PR https://github.com/vllm-project/vllm/pull/16367.

I suspect you might have used EAGLE-2 in SGLang, which explains the larger acceptance length for k = 4. Can you double check on that?

I will include the acceptance length comparison of vLLM against the EAGLE-1 repo soon. 

Regardless, I think we should merge this PR soon to unblock further developments. Without this PR, we can't even debug : )

cc @LiuXiaoxuanPKU @WoosukKwon

LiuXiaoxuanPKU committed:
force tokenizer

Signed-off-by: LiuXiaoxuanPKU <lilyliupku@gmail.com>

ekagra-ranjan commented:
@luyuzhe111 - thank you for your response. In SGL, I am using chain based draft where  `--speculative-num-steps 4 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4`. The diff bw EAGLE-1 and 2 is that draft tree is dynamic in EAGLE 2 but I am using a chain draft so I think it should be same as  EAGLE-1.

luyuzhe111 commented:
For reference, these are the acceptance length comparisons between EAGLE repo, vLLM v0, and vLLM v1. 

On [MT Bench](https://github.com/SafeAILab/EAGLE/blob/main/eagle/data/mt_bench/question.jsonl), assuming single-draft. Acceptance length is computed using https://github.com/vllm-project/vllm/pull/16367

When max number generated tokens = 256

| Number of Speculated Tokens | 1 | 2 | 3 | 4 | 5 |
|----------------------------|------|------|------|------|------|
| EAGLE Repo | 1.64 | 2.0 | 2.14 | 2.25 | 2.28 |
| vLLM v0 | 1.60 | 1.88 | 1.99 | 2.04 | 2.06 |
| vLLM v1 | 1.60 | 1.90 | 2.04 | 2.10 | 2.13 |

When max number generated tokens = 512

| Number of Speculated Tokens | 1 | 2 | 3 | 4 | 5 |
|----------------------------|------|------|------|------|------|
| EAGLE Repo | 1.65 | 2.01 | 2.19 | 2.28 | 2.33 |
| vLLM v0 | 1.61 | 1.87 | 1.99 | 2.04 | 2.06 |
| vLLM v1 | 1.61 | 1.91 | 2.05 | 2.11 | 2.14 |

Observations:
1. EAGLE implementation in v1 has better acceptance length than that in v0! Kudos to all the efforts to fix previously reported bugs!
2. The acceptance length for vLLM v1 EAGLE is still a bit lower than that from the EAGLE repo. Hopefully https://github.com/vllm-project/vllm/pull/16370 will bridge the gap.
3. In general, longer generation horizon gives better acceptance length. Intuitively, it's easier for EAGLE to speculate when there is more context from the base model. In the original EAGLE repo, the speculator performs slightly better when generation length increases from 256 to 512. This is not the case for v0 EAGLE, but we do observe a tiny improvement for v1 EAGLE. 

Hope the numbers here from the EAGLE repo can serve as a reference for future development efforts. cc @LiuXiaoxuanPKU @WoosukKwon

LiuXiaoxuanPKU committed:
Merge branch 'main' into eagle-weight

wwl2755 commented:
HI @luyuzhe111 , Great work! Nice to see such an early benchmarking! 

> 2\. The acceptance length for vLLM v1 EAGLE is still a bit lower than that from the EAGLE repo. Hopefully [[V1][Spec Decode] KV cache slots for eagle heads #16370](https://github.com/vllm-project/vllm/pull/16370) will bridge the gap.

IIUC, this PR is intended to allocate cache slot for the draft model, which should not affect the acceptance rate?

If the implementation is good, I assume it should be more related to the sampling/rejection method. And there indeed has some gaps, like https://github.com/vllm-project/vllm/blob/main/vllm/v1/spec_decode/eagle.py#L219. Maybe could you double-check whether the default sampling parameter consistent with reported in the original EAGLE?

ekagra-ranjan commented:
> IIUC, this PR is intended to allocate cache slot for the draft model, which should not affect the acceptance rate?

I have the same understanding. Can someone please share why #16370 would improve correctness?

WoosukKwon reviewed (approved):
LGTM! Thanks for the PR!

luyuzhe111 commented:
@wwl2755 @ekagra-ranjan Hi Wenlong and Ekagra, thanks for the comments. I actually don't understand https://github.com/vllm-project/vllm/pull/16370 too well and was hoping to dive deeper. Maybe we can keep the discussion under that PR now that this PR is merged?

Regarding the hypothesis on acceptance mechanism, I don't think sampling parameter is the issue since I used greedy sampling for both EAGLE repo and vLLM.

ekagra-ranjan commented:
@luyuzhe111 Thanks for sharing your [observation](https://github.com/vllm-project/vllm/pull/16035#issuecomment-2791539142)!

The issue was that SGL uses `speculative-num-draft-tokens - 1` number of draft tokens so the numbers I got for K=2 in SGL actually is comparable to K=1 in vLLM and similarly K=4 in SGL corresponds to K=3 in vLLM. The numbers are not better aligned :)

Using 1 draft token SGL gets 1.72 AL
Using 3 draft token SGL gets 2.4 AL
